### PR TITLE
Enhance Thread and Fiber Safety in ActiveRecordExaminer

### LINKED ISSUE #349: Add compatibility with Fiber-safe ConnectionPool in Rails 7.1
Rails 7.1 added opt-in `config.active_support.isolation_level = :fiber` in https://github.com/rails/rails/pull/44219 to replace Threads with more modern Fiber-based concurrency primitives that are available in 3.x versions of Ruby, but unfortunately appmap seems to cause conflicts with this setting that prevents database connections from being released to the pool, resulting in:

```
could not obtain a connection from the pool within 5.000 seconds (waited 5.004 seconds); all pooled connections were in use
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:128:in `block in wait_poll'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:117:in `loop'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:117:in `wait_poll'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:78:in `internal_poll'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:204:in `internal_poll'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:73:in `block in poll'
/opt/homebrew/Cellar/ruby/3.2.2_1/lib/ruby/3.2.0/monitor.rb:202:in `synchronize'
/opt/homebrew/Cellar/ruby/3.2.2_1/lib/ruby/3.2.0/monitor.rb:202:in `mon_synchronize'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:82:in `synchronize'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool/queue.rb:73:in `poll'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:661:in `acquire_connection'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:355:in `checkout'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:184:in `connection'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_handler.rb:246:in `retrieve_connection'
gems/activerecord-7.1.2/lib/active_record/connection_handling.rb:287:in `retrieve_connection'
gems/activerecord-7.1.2/lib/active_record/connection_handling.rb:254:in `connection'
gems/appmap-0.103.0/lib/appmap/handler/rails/sql_handler.rb:135:in `in_transaction?'
gems/appmap-0.103.0/lib/appmap/handler/rails/sql_handler.rb:49:in `examine'
gems/appmap-0.103.0/lib/appmap/handler/rails/sql_handler.rb:158:in `call'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:137:in `block in finish'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:26:in `block in iterate_guarding_exceptions'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:25:in `each'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:25:in `iterate_guarding_exceptions'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:125:in `each'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:136:in `finish'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:262:in `block in finish_with_values'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:26:in `block in iterate_guarding_exceptions'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:25:in `each'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:25:in `iterate_guarding_exceptions'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:261:in `finish_with_values'
gems/activesupport-7.1.2/lib/active_support/notifications/fanout.rb:254:in `finish'
gems/activesupport-7.1.2/lib/active_support/notifications/instrumenter.rb:64:in `instrument'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract_adapter.rb:1143:in `log'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:53:in `raw_execute'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/database_statements.rb:521:in `internal_execute'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/postgresql/schema_statements.rb:245:in `client_min_messages='
gems/activerecord-7.1.2/lib/active_record/connection_adapters/postgresql_adapter.rb:1001:in `configure_connection'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract_adapter.rb:744:in `reset!'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/postgresql_adapter.rb:377:in `block in reset!'
gems/activesupport-7.1.2/lib/active_support/concurrency/null_lock.rb:9:in `synchronize'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/postgresql_adapter.rb:369:in `reset!'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:426:in `block in reap'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:424:in `each'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:424:in `reap'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:660:in `acquire_connection'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:355:in `checkout'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_pool.rb:184:in `connection'
gems/activerecord-7.1.2/lib/active_record/connection_adapters/abstract/connection_handler.rb:246:in `retrieve_connection'
gems/activerecord-7.1.2/lib/active_record/connection_handling.rb:287:in `retrieve_connection'
gems/activerecord-7.1.2/lib/active_record/connection_handling.rb:254:in `connection'
gems/activerecord-7.1.2/lib/active_record/core.rb:362:in `cached_find_by_statement'
gems/activerecord-7.1.2/lib/active_record/core.rb:404:in `cached_find_by'
gems/activerecord-7.1.2/lib/active_record/core.rb:285:in `find_by'
gems/ar_lazy_preload-1.1.2/lib/ar_lazy_preload/active_record/base.rb:35:in `find_by'
gems/activerecord-7.1.2/lib/active_record/secure_password.rb:51:in `authenticate_by'
app/graphql/mutations/create_session.rb:45:in `authenticate_refresh_token'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:44:in `bind_call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:44:in `do_call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:53:in `trace_call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:20:in `call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:70:in `block in hook_method_def'
gems/skylight-6.0.1/lib/skylight/helpers.rb:184:in `authenticate_refresh_token'
app/graphql/mutations/create_session.rb:19:in `resolve'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:44:in `bind_call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:44:in `do_call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:53:in `trace_call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:20:in `call'
gems/appmap-0.103.0/lib/appmap/hook/method/ruby3.rb:70:in `block in hook_method_def'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/resolver.rb:107:in `public_send'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/resolver.rb:107:in `block (3 levels) in resolve_with_support'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:770:in `minimal_after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/query.rb:372:in `after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/resolver.rb:94:in `block (2 levels) in resolve_with_support'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:770:in `minimal_after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/query.rb:372:in `after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/resolver.rb:85:in `block in resolve_with_support'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:770:in `minimal_after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/query.rb:372:in `after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/resolver.rb:73:in `resolve_with_support'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/has_single_input_argument.rb:36:in `resolve_with_support'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/relay_classic_mutation.rb:45:in `resolve_with_support'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/field.rb:698:in `public_send'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/field.rb:698:in `block (2 levels) in resolve'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/field.rb:832:in `with_extensions'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/field.rb:669:in `block in resolve'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:770:in `minimal_after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/query.rb:372:in `after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/field.rb:667:in `resolve'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:358:in `block (2 levels) in evaluate_selection_with_resolved_keyword_args'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/tracing/trace.rb:52:in `execute_field'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:357:in `block in evaluate_selection_with_resolved_keyword_args'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:709:in `call_method_on_directives'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:354:in `evaluate_selection_with_resolved_keyword_args'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:331:in `block in evaluate_selection_with_args'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:830:in `after_lazy'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:279:in `evaluate_selection_with_args'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/runtime.rb:273:in `block in evaluate_selection'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/execution/interpreter/arguments_cache.rb:51:in `block in dataload_for'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/schema/member/has_arguments.rb:300:in `block (3 levels) in coerce_arguments'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/dataloader.rb:262:in `block in spawn_job_fiber'
bundler/gems/graphql-ruby-becd60008606/lib/graphql/dataloader.rb:230:in `block in spawn_fiber'
```

### PR DESCRIPTION
**Pull request by coderberry**
Attempt to fix #349 

~This PR fixes this issue by wrapping the Fiber that is accessing the connection pool with a temporary "root" Fiber. Once the query has completed, the fiber will be terminated and the connection will be returned to the pool.~

This PR is at a state where a fix can be applied and tested, but the implemented fix is not correct.

### CODE REVIEW

**Comment by coderberry** [comment]
@swrobel Would you mind reviewing this PR and provide any feedback or insights? I learned a _LOT_ from following you around in GitHub issues and discussions.

File: lib/appmap/handler/rails/sql_handler.rb [hunk 1]
```diff
@@ -132,16 +132,42 @@ def database_type
...
+                  ActiveRecord::Base.connection.execute(sql).to_a
+                end
+              end
+            end
+
+            private
+
+            # Assure Thread and Fiber safety for ActiveRecord queries.
+            # This is only necessary for applications using graphql-ruby.
+            # This change wraps the
```
Comment (coderberry) [inline lib/appmap/handler/rails/sql_handler.rb:154]: Will remove

File: lib/appmap/handler/rails/sql_handler.rb [hunk 2]
```diff
@@ -132,16 +132,42 @@ def database_type
...
+                  ActiveRecord::Base.connection_pool.with_connection { yield }
+                end
+              end
             end
           end
         end
 
-        def call(_, started, finished, _, payload) # (name, started, finished, unique_id, payload)
+        def call(_, started, finished, _, payload)
+          # (name, started, finished, unique_id, payload)
```
Comment (coderberry) [inline lib/appmap/handler/rails/sql_handler.rb:170]: Will correct

File: lib/appmap/handler/rails/sql_handler.rb (hunk 1 above)
Comment (dividedmind) [reply lib/appmap/handler/rails/sql_handler.rb:154]: Or finish the comment? :)

**Review by dividedmind — CHANGES_REQUESTED** [review]
The PR summary isn't informative; it reads as if it's machine-generated, it doesn't explain why the changes are necessary or correct. In fact I suspect that they're incorrect or at least inefficient. Could you please explain? I don't understand how that does anything to "improve fiber and thread safety" or indeed what problem is being solved.

EDIT: Sorry, I just read the linked issue, I see the problem. Still, with thread isolation, it seems the change cause a thread to be launched and two connections checked out from the pool for every query or `in_transaction?`; these threads then proceed to be zombies because they're never joined. This doesn't seem right, especially if there's already a thread being launched by ActiveSupport. Even in the fiber case, some tests would be useful to illustrate.

**Comment by coderberry** [comment]
@dividedmind thank you for your feedback. I forgot to flag this PR as a draft. Tests and a better description are coming 🙌🏻

File: lib/appmap/handler/rails/sql_handler.rb [hunk 3]
```diff
@@ -132,16 +132,42 @@ def database_type
...
+                end
+              end
+            end
+
+            private
+
+            # Assure Thread and Fiber safety for ActiveRecord queries.
+            # This is only necessary for applications using graphql-ruby.
+            # This change wraps the
+            def execute_in_context
```
Comment (swrobel) [inline lib/appmap/handler/rails/sql_handler.rb:155]: I believe [with_connection](https://github.com/rails/rails/pull/44219/files#diff-642b90553b888bd2c724c093a1a685a5408a7d8293f3751366c25dc548936eb7R210) already handles thread/fiber-safety, so I'm not sure if this conditional is required. Are you certain that it is?

Comment (coderberry) [reply lib/appmap/handler/rails/sql_handler.rb:155]: @swrobel it does, except in AppMap's case. I updated the description to explain why the wrapped Fiber is needed.

File: .nvmrc [hunk 4]
```diff
@@ -1 +1 @@
-14
+20.11.0
```
Comment (coderberry) [inline .nvmrc:1]: Will revert

File: lib/appmap/handler/rails/sql_handler.rb [hunk 5]
```diff
@@ -136,7 +136,32 @@ def in_transaction?
...
+            # wrapping the execution with `ActiveRecord::Base.connection_pool.with_connection`, we ensure the connection
+            # is properly released.
+            def execute_in_context
+              if defined?(ActiveSupport::IsolatedExecutionState) && ActiveSupport::IsolatedExecutionState.context == :fiber
+                # This is not working
+                Fiber.new do
+                  ActiveRecord::Base.connection_pool.with_connection { yield }
+                end.resume
+              else
+                ActiveRecord::Base.connection_pool.with_connection { |conn| block.call(conn) }
```
Comment (dividedmind) [inline lib/appmap/handler/rails/sql_handler.rb:163]: I see no `block` in this context.

**Pull request closed without merging**